### DISCUSSION TITLE
Replicating the huggingface repo

### DISCUSSION THREAD

**Discussion by antonyadhiban** [deepdoctection/deepdoctection#218] [General]
Hi @JaMe76, 
First I would like to thank you for building this amazing repo. 

**Question**: I would like to understand how I can replicate the huggingface space for private use (for a business use case). 
**Context**: Our customer would like to replace manual checking of an FMCG product packaging (eg: Cheetos cover) for text information. So running it locally or as a huggingface space with GPU enabled would be ideal. 
**Good to have**: I understand that it currently runs on CPU and sub 30 seconds results are impressive. But is it possible to run the same space with GPU and achieve faster results ?
**What I have tried so far**: I was able to run deepdoctection on an AWS EC2 machine. But I believe some parts of what makes the huggingface space work so well (like DD_ADDONS, Fully trained model) are not available, so the results are a bit underwhelming.

I am new to github discussions so please let me know if this is a bit out of scope.

**Comment by JaMe76** [comment]
Thank you for your kind words.

Regarding your question concerning HF space (duplicating, changing hardware, etc), I am unable to provide an answer and I could only point to the docs available for HF Spaces in general.

You are right, the deepdoctection space has better private models and AWS Textract as powerful OCR engine but DD_ADDONS itself does not really add a lot of magic. It only has a commercial PDF mining tool that is only needed to reduce Textract costs and a few NMS steps thereafter. I do not even think that the NMS post processing does still add any value, leaving it for historical reasons.
To take it short, only the models really make the difference…

**Reply by antonyadhiban** [reply]
Understood.

The replication doubt was related to the models and environment variables not being available. But now its clear.

We are currently close to 98% match (minor letter and punctuation mistakes) with deepdoctection under very specific setup (scanned copy of the plastic wrapper at the correct angle). Will we be able to achieve better results by training it against FMCG products (will be we be able to achieve 100% match) ? 

If we have to use your private model which is highly trained, do you have a commercial model or plan to do so ? 

Attaching a reference image which might help you understand the context a bit better. One is perfectly scanned (which gets 98% match), the other one is not (<10% match).

| aligned, cropped and resized to <1500px  | scanned by test user |
| ---------------------- | ---------------------- |
| ![download (3)](https://github.com/deepdoctection/deepdoctection/assets/16449971/40a7870e-afe4-4ec5-862a-8b0faa15d01b) | ![download (4)](https://github.com/deepdoctection/deepdoctection/assets/16449971/38094ad2-b2fd-4b0c-990d-6562c69c9f65) |

Technical doubts:
1. Resolution seems to have huge impact, the ideal conditions for the images seems to be >600px and <1500px. Beyond 2000px, the model seems to capture the layout in the image, but I am not getting output for text or table. 
2. Some of our documents have rotated text (90, 180, 270) which is completely ignored. If we get the layout bbox as output, I can manually rotate the cropped bbox image until I get a value. (or please let me know if there's a smarter solution to this.)
![image](https://github.com/deepdoctection/deepdoctection/assets/16449971/e0767ee6-78b4-4dd2-bc5d-f968b2b1b056)

Again, thank you for taking your time and building this solution, we are able to solve some very real problems with customers through this.

**Reply by JaMe76** [reply]
I would not expect 100% accuracy. Maybe only if you have a very narrow use-case and totally model overfit.

Layout detection works on images > 1400px. When passing the image to the model it will be re-scaled with largest edge size not exceeding 1400px. 
Table cells will be predicted from cropped table ROI. Even though it will be re-scaled, it is possible that the table region is out of training data distribution. Regarding text, I don’t know because I do not know what Textract is doing internally.

Textract does recognize text vertically but not that reliable as horizontally. Rotating is possible, the problem is of course to detect, what region must be rotated. Right know there is not so much support for rotated text. 

That’s for all Open Source related question. 

For everything beyond that you can reach me out here:

https://www.linkedin.com/in/dr-janis-meyer
